blockchain/tx_pool.go

```diff
@@ -924,10 +933,116 @@ func (pool *TxPool) HandleTxMsg(txs types.Transactions) {
 	if pool.config.DenyRemoteTx {
 		return
 	}
+
+	// Filter spam txs based on to-address of failed txs
+	spamThrottler := GetSpamThrottler()
+	if spamThrottler != nil {
+		pool.mu.RLock()
+		poolSize := uint64(len(pool.all))
+		pool.mu.RUnlock()
+
+		// Activate spam throttler when pool has enough txs
+		if poolSize > uint64(spamThrottler.config.ActivateTxPoolSize) {
+			allowTxs, throttleTxs := spamThrottler.classifyTxs(txs)
+
+			for _, tx := range throttleTxs {
+				select {
+				case spamThrottler.throttleCh <- tx:
+				default:
+					logger.Trace("drop a tx when throttleTxs channel is full", "txHash", tx.Hash())
+					throttlerDropCount.Inc(1)
+				}
+			}
+
+			txs = allowTxs
+		}
+	}
+
+	// TODO-Klaytn: Consider removing the next line and move the above logic to `addTx` or `AddRemotes`
 	senderCacher.recover(pool.signer, txs)
 	pool.txMsgCh <- txs
 }
 
+func (pool *TxPool) throttleLoop(spamThrottler *throttler) {
+	ticker := time.Tick(time.Second)
+	throttleNum := int(spamThrottler.config.ThrottleTPS)
+
+	for {
+		select {
+		case <-spamThrottler.quitCh:
+			logger.Info("Stop spam throttler loop")
+			return
+
+		case <-ticker:
+			txs := types.Transactions{}
+
+			iterNum := len(spamThrottler.throttleCh)
+			if iterNum > throttleNum {
+				iterNum = throttleNum
+			}
+
+			for i := 0; i < iterNum; i++ {
+				tx := <-spamThrottler.throttleCh
+				txs = append(txs, tx)
+			}
+
+			if len(txs) > 0 {
+				pool.AddRemotes(txs)
+			}
+		}
+	}
+}
+
+func (pool *TxPool) StartSpamThrottler(conf *ThrottlerConfig) error {
+	spamThrottlerMu.Lock()
+	defer spamThrottlerMu.Unlock()
+
+	if spamThrottler != nil {
+		return errors.New("spam throttler was already running")
+	}
+
+	if conf == nil {
+		conf = DefaultSpamThrottlerConfig
+	}
+
+	if err := validateConfig(conf); err != nil {
+		return err
+	}
+
+	t := &throttler{
+		config:     conf,
+		candidates: make(map[common.Address]int),
+		throttled:  make(map[common.Address]int),
+		allowed:    make(map[common.Address]bool),
+		mu:         new(sync.RWMutex),
+		threshold:  conf.InitialThreshold,
+		throttleCh: make(chan *types.Transaction, conf.ThrottleTPS*5),
+		quitCh:     make(chan struct{}),
+	}
+
+	go pool.throttleLoop(t)
+
+	spamThrottler = t
+	logger.Info("Start spam throttler", "config", *conf)
+	return nil
+}
+
+func (pool *TxPool) StopSpamThrottler() {
+	spamThrottlerMu.Lock()
```

Comment:
Don't we need any wait group?

Reply:
I thought using methods of the spamThrottler instance is okay after `spamThrottler =nil`. `StopSpamThrottler` doesn't free a throttler, but GC does.